### PR TITLE
[unittests] Add a fixture for Sema unit tests

### PR DESCRIPTION
**Pull request by xedin**
Adds a `SemaTest` fixture which handles all book-keeping related to setting up
environment and a simple literal binding inference test to demonstrate that
everything works correctly.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: unittests/Sema/SemaFixture.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  DiagnosticEngine Diags;
+
+  SemaTestBase() : Diags(SourceMgr) {
+    LangOpts.Target = llvm::Triple(llvm::sys::getProcessTriple());
+
+    std::string libDir = SWIFTLIB_DIR;
+    std::string platform = getPlatformNameForTriple(LangOpts.Target).str();
+
+    SearchPathOpts.RuntimeResourcePath = libDir;
+    SearchPathOpts.RuntimeLibraryPaths.push_back(libDir + "/" + platform);
```
Comment (artemcm) [inline unittests/Sema/SemaFixture.h:45]: This path separator will cause you trouble on Windows. LLVM must have a library for this somewhere, I am sure...

**Review by artemcm — APPROVED** [review]
Very cool!

File: unittests/Sema/SemaFixture.h (hunk 1 above)
Comment (xedin) [reply unittests/Sema/SemaFixture.h:45]: I'll take a look!

Comment (xedin) [reply unittests/Sema/SemaFixture.h]: Fixed!

**Comment by xedin** [comment]
@swift-ci please clean smoke test

**Comment by xedin** [comment]
@swift-ci please clean test Windows platform

**Comment by xedin** [comment]
@swift-ci please clean test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@compnerd Does Windows have `<build>/lib/swift` directory with built stdlib and other modules or artifacts are placed differently there? I'd hate it if I had to limit this tests to macOS only but I have to way to figure this out on my own...

File: unittests/Sema/SemaFixture.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  SearchPathOptions SearchPathOpts;
+  ClangImporterOptions ClangImporterOpts;
+  SourceManager SourceMgr;
+  DiagnosticEngine Diags;
+
+  SemaTestBase() : Diags(SourceMgr) {
+    LangOpts.Target = llvm::Triple(llvm::sys::getProcessTriple());
+
+    llvm::SmallString<128> libDir(SWIFTLIB_DIR);
+    llvm::sys::path::append(libDir, getPlatformNameForTriple(LangOpts.Target));
```
Comment (xedin) [inline unittests/Sema/SemaFixture.h:44]: I'm guessing that this doesn't work on Windows same way it does on macOS where stdlib module is located in `lib/swift/macosx`?

Comment (artemcm) [reply unittests/Sema/SemaFixture.h:44]: There is a piece of code in `lit.cfg` that computes the stdlib path (including on Windows) that seem to work:
https://github.com/apple/swift/blob/main/test/lit.cfg#L1495

**Comment by xedin** [comment]
@swift-ci please test Windows platform

File: unittests/Sema/SemaFixture.h (hunk 2 above)
Comment (compnerd) [reply unittests/Sema/SemaFixture.h:44]: The layout on Windows is similar to that on macOS: `lib/swift/windows/Swift.swiftmodule/x86_64-unknown-windows-msvc.swiftmodule`.

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test macOS platform

File: unittests/Sema/SemaFixture.h (hunk 2 above)
Comment (xedin) [reply unittests/Sema/SemaFixture.h:44]: Thanks to @compnerd, we have figured that I should have used default target instead. Now everything is fine.

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Pull request merged**